internal/mempool/v1/mempool.go

```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
-// inserted into the txStore and indexed across all indexes. Otherwise, if the
-// mempool is full, we attempt to find a lower priority transaction to evict in
-// place of the new incoming transaction. If no such transaction exists, the
-// new incoming transaction is rejected.
+// If either the application rejected the transaction or a post-check hook is
+// defined and rejects the transaction, it is discarded.
 //
-// If the new incoming transaction fails CheckTx or postCheck fails, we reject
-// the new incoming transaction.
+// Otherwise, if the mempool is full, check for lower-priority transactions
+// that can be evicted to make room for the new one. If no such transactions
+// exist, this transaction is logged and dropped; otherwise the selected
+// transactions are evicted.
 //
-// NOTE:
-// - An explicit lock is NOT required.
-func (txmp *TxMempool) initTxCallback(wtx *WrappedTx, res *abci.Response, txInfo mempool.TxInfo) {
+// Finally, the new transaction is added and size stats updated.
+func (txmp *TxMempool) initTxCallback(wtx *WrappedTx, res *abci.Response) {
 	checkTxRes, ok := res.Value.(*abci.Response_CheckTx)
 	if !ok {
 		return
 	}
 
+	txmp.mtx.Lock()
+	defer txmp.mtx.Unlock()
+
 	var err error
 	if txmp.postCheck != nil {
 		err = txmp.postCheck(wtx.tx, checkTxRes.CheckTx)
 	}
 
 	if err != nil || checkTxRes.CheckTx.Code != abci.CodeTypeOK {
-		// ignore bad transactions
 		txmp.logger.Info(
 			"rejected bad transaction",
-			"priority", wtx.priority,
+			"priority", wtx.Priority(),
 			"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
-			"peer_id", txInfo.SenderNodeID,
+			"peer_id", wtx.peers,
 			"code", checkTxRes.CheckTx.Code,
 			"post_check_err", err,
 		)
 
 		txmp.metrics.FailedTxs.Add(1)
 
+		// Remove the invalid transaction from the cache, unless the operator has
+		// instructed us to keep invalid transactions.
 		if !txmp.config.KeepInvalidTxsInCache {
 			txmp.cache.Remove(wtx.tx)
 		}
+
+		// If there was a post-check error, record its text in the result for
+		// debugging purposes.
 		if err != nil {
 			checkTxRes.CheckTx.MempoolError = err.Error()
 		}
 		return
 	}
 
-	sender := checkTxRes.CheckTx.Sender
 	priority := checkTxRes.CheckTx.Priority
+	sender := checkTxRes.CheckTx.Sender
 
-	if len(sender) > 0 {
-		if wtx := txmp.txStore.GetTxBySender(sender); wtx != nil {
+	// Disallow multiple concurrent transactions from the same sender assigned
+	// by the ABCI application. As a special case, an empty sender is not
+	// restricted.
+	if sender != "" {
+		elt, ok := txmp.txBySender[sender]
+		if ok {
+			w := elt.Value.(*WrappedTx)
 			txmp.logger.Error(
-				"rejected incoming good transaction; tx already exists for sender",
-				"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
+				"rejected valid incoming transaction; tx already exists for sender",
+				"tx", fmt.Sprintf("%X", w.tx.Hash()),
 				"sender", sender,
 			)
-			txmp.metrics.RejectedTxs.Add(1)
 			return
 		}
```

Comment:
Can we append this error to `checkTxRes.CheckTx.MempoolError` so RPC callers understand that their tx failed and why